lucene/core/src/java/org/apache/lucene/search/MaxScoreTerminator.java

```diff
@@ -0,0 +1,170 @@
+  int totalCollected;
+
+  /** the worst hit over all */
+  LeafState leafState;
+
+  MaxScoreTerminator(int totalToCollect) {
+    leafStates = new ArrayList<>();
+    this.totalToCollect = totalToCollect;
+    setIntervalBits(DEFAULT_INTERVAL_BITS);
+  }
+
+  synchronized LeafState add() {
+    LeafState newLeafState = new LeafState();
+    leafStates.add(newLeafState);
+    if (leafState == null) {
+      leafState = newLeafState;
+    }
+    return newLeafState;
+  }
+
+  // for testing
+  void setIntervalBits(int bitCount) {
+    interval = 1 << bitCount;
+    intervalMask = interval - 1;
+  }
+
+  /**
+   * Called by leaf collectors periodically to update their progress.
+   * @param newLeafState the leaf collector's current lowest score
+   * @return whether the collector should terminate
+   */
+  synchronized boolean update(LeafState newLeafState) {
+    totalCollected += interval;
+    //System.out.println(" scoreboard totalCollected = " + totalCollected + "/" + totalToCollect + " "
+    //  + newLeafState + " ? " + leafState + ":" + newLeafState.compareTo(leafState));
+    if (newLeafState.compareTo(leafState) >= 0 || newLeafState.resultCount >= totalToCollect) {
+      leafState = newLeafState;
+      if (totalCollected >= totalToCollect) {
+        // in this case, we have enough scores: remove the worst leaves
+        updateWorstHit();
+        // and tell the current leaf collector to terminate since it can
+        // no longer contribute any top hits
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private void updateWorstHit() {
+    //System.out.println("scoreboard updateWorstHit");
+    if (leafStates.size() > 1) {
+      Collections.sort(leafStates);
+      while (leafStates.size() > 1) {
+        LeafState worst = leafStates.get(leafStates.size() - 1);
+        if (totalCollected - numExcludedBound - worst.resultCount >= totalToCollect) {
+          //System.out.println(" and remove");
+          // We don't need the worst leaf in order to get enough results, so
+          // remember how many results (upper bound) it accounted for
+          numExcludedBound += worst.resultCount;
+          // and remove it from the list of worstHits
+          leafStates.remove(leafStates.size() - 1);
+        } else {
+          break;
+        }
+      }
+      // and update the worstHit with the worst one that remains after exclusion
+      leafState = leafStates.get(leafStates.size() - 1);
+    }
+  }
+
+  /**
+   * For tracking the worst scoring hit and total number of hits collected by each leaf
+   */
+  class LeafState implements Comparable<LeafState> {
+
+    private float worstScore;
+    int resultCount;
+
+    LeafState() {
+      this.worstScore = -1;
+    }
+
+    void update(float score) {
+      assert score >= this.worstScore;
+      this.worstScore = score;
+      ++this.resultCount;
+    }
+
+    @Override
+    public int compareTo(LeafState o) {
+      return Float.compare(worstScore, o.worstScore);
```

Comment:
Hmm why `Float` and not `Double`?  What if the sort was a compound sort (sorting on multiple fields)?

Reply:
We only terminate based on the first component of the score; we basically behave as if subsequent components were all equal. Hmm could be a lurking bug there. We might have to limit this optimization to Sort where this only a single field. Given that, is float sufficient? I guess we could be sorting on a double-valued field, so we should switch to double.

Reply:
OK multiple fields is not a problem. After a long detour (where I learned how our test seeds work), I found and fixed a small bug when we use multiple sort keys with inverse ordering. Also, I am posting a new commit that switches to using double for scores.